cpp/src/genetic/genetic.cu

```diff
@@ -14,13 +14,515 @@
+      }
+    } 
+
+    // Run tournaments
+    rmm::device_uvector<uint64_t> tour_seeds(n_tours,stream);
+    rmm::device_uvector<int> d_win_indices(n_tours,stream);
+    d_gen.uniformInt(tour_seeds.data(), n_tours, (uint64_t)1, (uint64_t)INT_MAX, stream);
+
+    auto crit = params.criterion();
+    dim3 nblks(raft::ceildiv(n_tours,GENE_TPB),1,1);
+    batched_tournament_kernel<<<nblks,GENE_TPB,0,stream>>>(d_oldprogs,
+                                                           d_win_indices.data(),tour_seeds.data(),
+                                                           n_progs,n_tours,
+                                                           tour_size,crit);
+    CUDA_CHECK(cudaPeekAtLastError());
+
+    // Make sure tournaments have finished running before copying win indices
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+    
+    // Perform host mutations
+
+    auto donor_pos  = n_progs;
+    for(auto pos=0; pos < n_progs; ++pos) {
+      
+      auto parent_index = d_win_indices.element(pos, stream);
+      
+      if(h_nextprogs[pos].mut_type == mutation_t::crossover){
+        // Get secondary index
+        auto donor_index = d_win_indices.element(donor_pos, stream);
+        donor_pos++; 
+        crossover(h_oldprogs[parent_index], h_oldprogs[donor_index],h_nextprogs[pos], params, h_gen);
+      }
+      else if(h_nextprogs[pos].mut_type == mutation_t::subtree){
+        subtree_mutation(h_oldprogs[parent_index],h_nextprogs[pos],params, h_gen);
+      }
+      else if(h_nextprogs[pos].mut_type == mutation_t::hoist){
+        hoist_mutation(h_oldprogs[parent_index],h_nextprogs[pos],params,h_gen);
+      }
+      else if(h_nextprogs[pos].mut_type == mutation_t::point){
+        point_mutation(h_oldprogs[parent_index],h_nextprogs[pos],params,h_gen);
+      }
+      else if(h_nextprogs[pos].mut_type == mutation_t::reproduce){
+        h_nextprogs[pos] = h_oldprogs[parent_index];
+      }
+      else{
+        // Should not come here
+      }
+    }
+  }
+
+  /* Memcpy individual host nodes to device and destroy previous generation device nodes
+     TODO: Find a better way to do this. */
+  for(auto i=0;i<n_progs;++i) {
+
+    program tmp(h_nextprogs[i]);        
+    delete[] tmp.nodes;
+
+    // Set current generation device nodes
+    tmp.nodes = (node*)h.get_device_allocator()->allocate(h_nextprogs[i].len*sizeof(node),stream);
+    CUDA_CHECK(cudaMemcpyAsync(tmp.nodes, h_nextprogs[i].nodes,h_nextprogs[i].len * sizeof(node),
+                               cudaMemcpyHostToDevice, stream));
+    CUDA_CHECK(cudaMemcpyAsync(d_nextprogs + i, &tmp,sizeof(program),cudaMemcpyHostToDevice,stream));
+
+    if(generation > 1){
+      // Free device memory allocated to program nodes in previous generation
+      CUDA_CHECK(cudaMemcpyAsync(&tmp,d_oldprogs + i,sizeof(program),cudaMemcpyDeviceToHost,stream));
+      h.get_device_allocator()->deallocate(tmp.nodes,h_oldprogs[i].len*sizeof(node),stream);
+    }
+    
+    tmp.nodes = nullptr;
```

Comment:
```suggestion
    CUDA_CHECK(cudaMemcpyAsync(d_nextprogs + i, &h_nextprogs[i], sizeof(program), cudaMemcpyHostToDevice, stream));
    (d_nextprogs + i)->nodes = (node*)h.get_device_allocator()->allocate(h_nextprogs[i].len*sizeof(node),stream);
    CUDA_CHECK(cudaMemcpyAsync((d_nextprogs + i)->nodes, h_nextprogs[i].nodes, h_nextprogs[i].len * sizeof(node),
                               cudaMemcpyHostToDevice, stream));
    if(generation > 1){
      h.get_device_allocator()->deallocate((d_oldprogs + i)->nodes,h_oldprogs[i].len*sizeof(node),stream);
    }
```
This removes the middleman `tmp` and possibly the redundant allocation (when you do `program tmp(h_nextprogs[i])` and then `delete[] tmp.nodes` in the very next line). Just a nitpick though.

Reply:
@venkywonka I have already optimized this from 3 to 2 `memcpyAsync` calls.  `program tmp` is kind of like a necessary evil here :)
Trying to allocate memory to `(d_nextprogs+i)->nodes` should give an segfault (and it does), as when the allocator eventually does `cudaMalloc(&((d_nextprogs+i)->nodes),...)`, `&((d_nextprogs+i)->nodes)` is expected to reside in host memory, which isn't the case here.  
The snippet below illustrates the code without `program tmp`. Note the need for 3 `memcpyAsync` calls!  
```
    node* tmp_nodes;

    // Set current generation device nodes
    tmp_nodes = (node*)h.get_device_allocator()->allocate(h_nextprogs[i].len*sizeof(node),stream);
    
    CUDA_CHECK(cudaMemcpyAsync(d_nextprogs + i, &tmp,sizeof(program),cudaMemcpyHostToDevice,stream));
    CUDA_CHECK(cudaMemcpyAsync(tmp_nodes, h_nextprogs[i].nodes,h_nextprogs[i].len * sizeof(node),
                               cudaMemcpyHostToDevice, stream));
    CUDA_CHECK(cudaMemcpyAsync(&((d_nextprogs+i)->nodes),&tmp_nodes,sizeof(node*),cudaMemcpyHostToDevice,stream));

    if(generation > 1){
      // Free device memory allocated to program nodes in previous generation
      CUDA_CHECK(cudaMemcpyAsync(&tmp_nodes,&((d_oldprogs + i)->nodes),sizeof(node*),cudaMemcpyDeviceToHost,stream));
      h.get_device_allocator()->deallocate(tmp_nodes,h_oldprogs[i].len*sizeof(node),stream);
    }
    
    tmp_nodes = nullptr;
```

Reply:
@venkywonka It is already optimized!
Directly allocating memory to `(d_nextprogs+i)->nodes` should lead to a segfault(and it does, as `&((d_nextprogs+i)->nodes)` is expected to be a host memory location, for the internal `cudaMallocd` pointer to be properly copied to `&((d_nextprogs+i)->nodes)`).

The only way is to either use a `node*` temporary array or a `program` tmp object(Necessary evil here due to the deepcopy problem).

However, using a `node*` temporary array will take up 3 `memcpyAsync` calls instead of 2!
```
    node* tmp_nodes;

    // Set current generation device nodes
    tmp_nodes = (node*)h.get_device_allocator()->allocate(h_nextprogs[i].len*sizeof(node),stream);
    CUDA_CHECK(cudaMemcpyAsync(d_nextprogs + i, &tmp,sizeof(program),cudaMemcpyHostToDevice,stream));
    CUDA_CHECK(cudaMemcpyAsync(tmp_nodes, h_nextprogs[i].nodes,h_nextprogs[i].len * sizeof(node),
                               cudaMemcpyHostToDevice, stream));
    CUDA_CHECK(cudaMemcpyAsync(&((d_nextprogs+i)->nodes),&tmp_nodes,sizeof(node*),cudaMemcpuHostToDevice,stream));

    if(generation > 1){
      // Free device memory allocated to program nodes in previous generation
      CUDA_CHECK(cudaMemcpyAsync(&tmp_nodes,&(d_oldprogs + i)->nodes,sizeof(node*),cudaMemcpyDeviceToHost,stream));
      h.get_device_allocator()->deallocate(tmp_nodes,h_oldprogs[i].len*sizeof(node),stream);
    }
    
    tmp_nodes = nullptr;
```

Reply:
oh! that's weird, I tried compiling and running the tests, and it seemed to work 🤔 are you sure it would lead to a segfault?
If so, how did you reproduce the error?
It didn't seem to when I did `./test/ml --gtest_filter=Genetic*` (maybe I'm not covering all the tests?)

Reply:
Yeah, GoogleTest only has tests for programs + fitness functions. There's no test for the evolution of a whole generation. Should I add one?

Reply:
You're right, didn't notice the constraint that would cause segfault! Thanks for adding the tests vimarsh, now feels more complete 😄